Insert custom code in the autogenerated PackageConfig.cmake file

diegoferigo opened the pull request:
This commit adds to `InstallBasicPackageFiles.cmake` the support of inserting custom code read from a file to the autogenerated `PackageConfig.cmake`.

This can be useful e.g. for setting variables such as `CMAKE_MODULE_PATH` to automatically load exported macros and functions. Under the hood `configure_file` is used. This gives the possibility of specifying a templated input file.

In the mentioned example, here is a test use case:

```cmake
# CustomCodeToInsert.cmake.in
list(APPEND CMAKE_MODULE_PATH @CMAKE_INSTALL_PREFIX@/share/Foo/cmake)
```

```cmake
# Upstream project
include(InstallBasicPackageFiles)
install_basic_package_files(ExportedPackageName
    VERSION ${PROJECT_VERSION}
    DEPENDENCIES ${ExportedPackageName_DEPENDENCIES}
    COMPATIBILITY AnyNewerVersion
    TARGETS_PROPERTY ExportedPackageName_TARGETS
    NO_CHECK_REQUIRED_COMPONENTS_MACRO
    INCLUDE_FILE ${CMAKE_SOURCE_DIR}/cmake/CustomCodeToInsert.cmake.in
)
```

It would result in the following `ExportedPackageName.cmake`

```cmake
set(ExportedPackageName_VERSION 1.0)


####### Expanded from @PACKAGE_INIT@ by configure_package_config_file() #######
####### Any changes to this file will be overwritten by the next CMake run ####
####### The input file was ExportedPackageNameConfig.cmake.in                            ########

[...]

####################################################################################

#### Expanded from @PACKAGE_DEPENDENCIES@ by install_basic_package_files() ####

include(CMakeFindDependencyMacro)
find_dependency(...)

###############################################################################


#########################################################
#### Expanded from INCLUDE_FILE                      ####

list(APPEND CMAKE_MODULE_PATH /iit/sources/robotology-superbuild/build/install/share/Foo/cmake)

#########################################################

# Compatibility
[...]
```

Feedback on this is much appreciated @drdanz @traversaro.

drdanz commented on modules/InstallBasicPackageFiles.cmake:
```diff
@@ -405,6 +435,8 @@ function(INSTALL_BASIC_PACKAGE_FILES _Name)
 \@PACKAGE_DEPENDENCIES\@
 
+${_includedfile_content}
+
```
I think it would be better to include the extra content at the end of the file.

Also I'd rather generate a file that contains some "@-variable" that is expanded later, i.e. `\@INCLUDED_FILE_CONTENT\@`, and remove the `message(WARNING)`

This will allow to use this feature also with a custom file, probably not really useful in production, but it might come handy for making tests...

diegoferigo replied:
> and remove the message(WARNING)

What should be the behavior in the case the user specifies both `INCLUDE_FILE` and `CONFIG_TEMPLATE`?

drdanz replied:
It should read the include file, and save the content in an `INCLUDED_FILE_CONTENT` variable (exactly how it should do in the other case)

When the `CONFIG_TEMPLATE` is configured, the `@INCLUDED_FILE_CONTENT@` tag(s) in the file is (are) replaced by CMake like `@PACKAGE_DEPENDENCIES@`, `@PACKAGE_INIT@`...

drdanz commented on modules/InstallBasicPackageFiles.cmake:
```diff
@@ -348,6 +355,24 @@ function(INSTALL_BASIC_PACKAGE_FILES _Name)
+  # Set input file containing user variables
+  if(DEFINED _IBPF_INCLUDE_FILE)
+    if(NOT EXISTS ${_IBPF_INCLUDE_FILE})
+        message(FATAL_ERROR "File \"${_IBPF_INCLUDE_FILE}\" not found")
```
I did a very quick test, and I ended up immediately in this `FATAL_ERROR`.
I think you should check if the filename is absolute or not, and in the second case prepend `
`${CMAKE_CURRENT_SOURCE_DIR}`

drdanz commented on modules/InstallBasicPackageFiles.cmake:
```diff
@@ -348,6 +355,24 @@ function(INSTALL_BASIC_PACKAGE_FILES _Name)
+    file(READ ${_includedfile_outputname} _includedfile_user_content)
+    set(INCLUDED_FILE_CONTENT
+"#########################################################
+#### Expanded from INCLUDE_FILE                      ####")
```
```diff
-#### Expanded from INCLUDE_FILE                      ####"
+#### Expanded from @INCLUDED_FILE_CONTENT@ by install_basic_package_files() ####
```

drdanz commented on modules/InstallBasicPackageFiles.cmake:
```diff
@@ -348,6 +355,24 @@ function(INSTALL_BASIC_PACKAGE_FILES _Name)
+    get_filename_component(_includedfile_name ${_includedfile_in} NAME_WE)
+    set(_includedfile_outputname ${CMAKE_CURRENT_BINARY_DIR}/${_includedfile_name}.cmake)
+    configure_file(${_includedfile_in} ${_includedfile_outputname})
+    file(READ ${_includedfile_outputname} _includedfile_user_content)
```
you can use [`string(CONFIGURE)`](https://cmake.org/cmake/help/git-master/command/string.html#configure) here, no need to write the file...

diegoferigo replied:
Nice, I didn't know this feature

drdanz commented:
BTW, I tested it (it took way less than a couple of days :wink:) and it works great, great feature! Thanks!

drdanz commented:
Can you please add also a line to the release notes? `help/release/0.8.0.rst`

diegoferigo commented:
@drdanz Addressed all your comments, thanks! As soon as you approve the PR, I rebase it squashing the `[squash]` commits to get this ready to be merged.

traversaro commented on modules/InstallBasicPackageFiles.cmake:
```diff
@@ -348,6 +355,29 @@ function(INSTALL_BASIC_PACKAGE_FILES _Name)
+    #get_filename_component(_includedfile_name ${_includedfile_in} NAME_WE)
+    #set(_includedfile_outputname ${CMAKE_CURRENT_BINARY_DIR}/${_includedfile_name}.cmake)
+    #configure_file(${_includedfile_in} ${_includedfile_outputname})
+#    file(READ ${_includedfile_outputname} _includedfile_user_content)
```
I guess we can remove this commented lines if we go for the `string(CONFIGURE ..)` solution?

diegoferigo replied:
Of course :)

diegoferigo commented:
Ready

drdanz commented:
Merged, thanks.

drdanz commented:
Sorry, I didn't notice this PR was targeting master, I reverted this on `master` (force push) and cherry picked it on `devel`

diegoferigo commented:
Sorry, I didn't realize it, my mistake :/

traversaro commented:
See the following links for the variables to use for create relocatable packages:
* https://github.com/robotology/ycm/blob/767eceead4ce7db9040d4070ca98868fdbfe8057/modules/InstallBasicPackageFiles.cmake#L127
* `PATH_VARS` in https://cmake.org/cmake/help/v3.8/module/CMakePackageConfigHelpers.html
